File: llamafile/server/www/highlight_css.js
// Copyright 2024 Mozilla Foundation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

class HighlightCss extends Highlighter {

  static NORMAL = 0;
  static SELECTOR = 1;
  static PROPERTY = 2;
  static VALUE = 3;
  static QUOTE = 4;
  static QUOTE_BACKSLASH = 5;
  static DQUOTE = 6;
  static DQUOTE_BACKSLASH = 7;
  static SLASH = 8;
  static SLASH_STAR = 9;
  static SLASH_STAR_STAR = 10;

  constructor(delegate) {
    super(delegate);
    this.pushed = 0;
  }

  push(tagName, className) {
    while (this.pushed)
      this.pop();
    ++this.pushed;
    return this.delegate.push(tagName, className);
  }

  pop() {
    if (!this.pushed)
      throw new Error('bad pop');
    --this.pushed;
    this.delegate.pop();
  }

  feed(input) {
    for (let i = 0; i < input.length; i += this.delta) {
      this.delta = 1;
      let c = input[i];

      switch (this.state & 255) {

      case HighlightCss.NORMAL:
        this.push("span", "selector");
        this.state = HighlightCss.SELECTOR;
        // fallthrough

      case HighlightCss.SELECTOR:
        if (c == '{') {
          this.state = HighlightCss.PROPERTY;
          this.pop();
          this.append(c);
          this.push("span", "property");
        } else if (c == ',') {
          this.pop();
          this.append(c);
          this.push("span", "selector");
        } else if (c == '/') {
          this.state = HighlightCss.SELECTOR << 8 | HighlightCss.SLASH;
        } else if (c == '\'') {
          this.state = HighlightCss.SELECTOR << 8 | HighlightCss.QUOTE;
          this.push("span", "string");
          this.append(c);
        } else if (c == '"') {
          this.state = HighlightCss.SELECTOR << 8 | HighlightCss.DQUOTE;
          this.push("span", "string");
          this.append(c);
        } else {
          this.append(c);
        }
        break;

      case HighlightCss.PROPERTY:
        if (c == '/') {
          this.state = HighlightCss.PROPERTY << 8 | HighlightCss.SLASH;
        } else if (c == '\'') {
          this.state = HighlightCss.PROPERTY << 8 | HighlightCss.QUOTE;
          this.push("span", "string");
          this.append(c);
        } else if (c == '"') {
          this.state = HighlightCss.PROPERTY << 8 | HighlightCss.DQUOTE;
          this.push("span", "string");
          this.append(c);
        } else if (c == ':') {
          this.state = HighlightCss.VALUE;
          this.pop();
          this.append(c);
        } else if (c == '}') {
          this.state = HighlightCss.SELECTOR;
          this.pop();
          this.append(c);
          this.push("span", "selector");
        } else {
          this.append(c);
        }
        break;

      case HighlightCss.VALUE:
        if (c == '/') {
          this.state = HighlightCss.VALUE << 8 | HighlightCss.SLASH;
        } else if (c == '\'') {
          this.state = HighlightCss.VALUE << 8 | HighlightCss.QUOTE;
          this.push("span", "string");
          this.append(c);
        } else if (c == '"') {
          this.state = HighlightCss.VALUE << 8 | HighlightCss.DQUOTE;
          this.push("span", "string");
          this.append(c);
        } else if (c == ';') {
          this.state = HighlightCss.PROPERTY;
          this.append(c);
          this.push("span", "property");
        } else if (c == '}') {
          this.state = HighlightCss.SELECTOR;
          this.append(c);
          this.push("span", "selector");
        } else {
          this.append(c);
        }
        break;

      case HighlightCss.SLASH:
        if (c == '*') {
          this.push("span", "comment");
          this.append("/*");
          this.state &= -256;
          this.state |= HighlightCss.SLASH_STAR;
        } else {
          this.append('/');
          this.state >>= 8;
          this.epsilon(this.state);
        }
        break;

      case HighlightCss.SLASH_STAR:
        this.append(c);
        if (c == '*') {
          this.state &= -256;
          this.state |= HighlightCss.SLASH_STAR_STAR;
        }
        break;

      case HighlightCss.SLASH_STAR_STAR:
        this.append(c);
        if (c == '/') {
          this.pop();
          this.state >>= 8;
          if (this.state == HighlightCss.SELECTOR)
            this.push("span", "selector");
          if (this.state == HighlightCss.PROPERTY)
            this.push("span", "property");
        } else if (c != '*') {
          this.state &= -256;
          this.state |= HighlightCss.SLASH_STAR;
        }
        break;

      case HighlightCss.QUOTE:
        this.append(c);
        if (c == '\'') {
          this.pop();
          this.state >>= 8;
          if (this.state == HighlightCss.SELECTOR)
            this.push("span", "selector");
          if (this.state == HighlightCss.PROPERTY)
            this.push("span", "property");
        } else if (c == '\\') {
          this.state &= -256;
          this.state |= HighlightCss.QUOTE_BACKSLASH;
        }
        break;

      case HighlightCss.QUOTE_BACKSLASH:
        this.append(c);
        this.state &= -256;
        this.state |= HighlightCss.QUOTE;
        break;

      case HighlightCss.DQUOTE:
        this.append(c);
        if (c == '"') {
          this.pop();
          this.state >>= 8;
          if (this.state == HighlightCss.SELECTOR)
            this.push("span", "selector");
          if (this.state == HighlightCss.PROPERTY)
            this.push("span", "property");
        } else if (c == '\\') {
          this.state &= -256;
          this.state |= HighlightCss.DQUOTE_BACKSLASH;
        }
        break;

      case HighlightCss.DQUOTE_BACKSLASH:
        this.append(c);
        this.state &= -256;
        this.state |= HighlightCss.DQUOTE;
        break;

      default:
        throw new Error('Invalid state');
      }
    }
  }

  flush() {
    switch (this.state & 255) {
    case HighlightCss.SLASH:
      this.append('/');
      break;
    default:
      break;
    }
    while (this.pushed)
      this.pop();
    this.state = HighlightCss.NORMAL;
    this.delegate.flush();
    this.delta = 1;
  }
}

Highlighter.REGISTRY['css'] = HighlightCss;

File: llamafile/highlight/highlight_css.cpp
// -*- mode:c++;indent-tabs-mode:nil;c-basic-offset:4;coding:utf-8 -*-
// vi: set et ft=cpp ts=4 sts=4 sw=4 fenc=utf-8 :vi
//
// Copyright 2024 Mozilla Foundation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "highlight.h"

#include <ctype.h>

enum
{
    NORMAL,
    SELECTOR,
    PROPERTY,
    VALUE,
    QUOTE,
    QUOTE_BACKSLASH,
    DQUOTE,
    DQUOTE_BACKSLASH,
    SLASH,
    SLASH_STAR,
    SLASH_STAR_STAR,
    AT,
    BANG,
};

HighlightCss::HighlightCss()
{
}

HighlightCss::~HighlightCss()
{
}

void
HighlightCss::feed(std::string* r, std::string_view input)
{
    int c;
    for (size_t i = 0; i < input.size(); ++i) {
        c = input[i] & 255;

    TryAgain:
        switch (t_ & 255) {

        case NORMAL:
            *r += HI_SELECTOR;
            t_ = SELECTOR;
            // fallthrough

        Selector:
        case SELECTOR:
            if (c == '{') {
                t_ = PROPERTY;
                *r += HI_RESET;
                *r += c;
                *r += HI_PROPERTY;
            } else if (c == ',') {
                *r += HI_RESET;
                *r += c;
                *r += HI_SELECTOR;
            } else if (c == '/') {
                t_ = SELECTOR << 8 | SLASH;
            } else if (c == '@') {
                t_ = SELECTOR << 8 | AT;
            } else if (c == '\'') {
                t_ = SELECTOR << 8 | QUOTE;
                *r += HI_STRING;
                *r += c;
            } else if (c == '"') {
                t_ = SELECTOR << 8 | DQUOTE;
                *r += HI_STRING;
                *r += c;
            } else {
                *r += c;
            }
            break;

        Property:
        case PROPERTY:
            if (c == '/') {
                t_ = PROPERTY << 8 | SLASH;
            } else if (c == '@') {
                t_ = VALUE << 8 | AT;
            } else if (c == '\'') {
                t_ = PROPERTY << 8 | QUOTE;
                *r += HI_STRING;
                *r += c;
            } else if (c == '"') {
                t_ = PROPERTY << 8 | DQUOTE;
                *r += HI_STRING;
                *r += c;
            } else if (c == ':') {
                t_ = VALUE;
                *r += HI_RESET;
                *r += c;
            } else if (c == '}') {
                t_ = SELECTOR;
                *r += HI_RESET;
                *r += c;
                *r += HI_SELECTOR;
            } else {
                *r += c;
            }
            break;

        Value:
        case VALUE:
            if (c == '/') {
                t_ = VALUE << 8 | SLASH;
            } else if (c == '@') {
                t_ = VALUE << 8 | AT;
            } else if (c == '!') {
                t_ = VALUE << 8 | BANG;
            } else if (c == '\'') {
                t_ = VALUE << 8 | QUOTE;
                *r += HI_STRING;
                *r += c;
            } else if (c == '"') {
                t_ = VALUE << 8 | DQUOTE;
                *r += HI_STRING;
                *r += c;
            } else if (c == ';') {
                t_ = PROPERTY;
                *r += c;
                *r += HI_PROPERTY;
            } else if (c == '}') {
                t_ = SELECTOR;
                *r += c;
                *r += HI_SELECTOR;
            } else {
                *r += c;
            }
            break;

        case AT:
            if (isalpha(c) || c == '-') {
                word_ += c;
            } else {
                if (is_keyword_css_at(word_.data(), word_.size())) {
                    *r += HI_BUILTIN;
                    *r += '@';
                    *r += word_;
                    *r += HI_RESET;
                } else {
                    *r += '@';
                    *r += word_;
                }
                word_.clear();
                t_ >>= 8;
                switch (t_ & 255) {
                case SELECTOR:
                    *r += HI_SELECTOR;
                    goto Selector;
                case PROPERTY:
                    *r += HI_PROPERTY;
                    goto Property;
                case VALUE:
                    goto Value;
                default:
                    __builtin_unreachable();
                }
            }
            break;

        case BANG:
            if (isalpha(c) || c == '-') {
                word_ += c;
            } else {
                if (is_keyword_css_bang(word_.data(), word_.size())) {
                    *r += HI_WARNING;
                    *r += '!';
                    *r += word_;
                    *r += HI_RESET;
                } else {
                    *r += '!';
                    *r += word_;
                }
                word_.clear();
                t_ >>= 8;
                switch (t_ & 255) {
                case VALUE:
                    goto Value;
                default:
                    __builtin_unreachable();
                }
            }
            break;

        case SLASH:
            if (c == '*') {
                *r += HI_COMMENT;
                *r += "/*";
                t_ &= -256;
                t_ |= SLASH_STAR;
            } else {
                *r += '/';
                t_ >>= 8;
                goto TryAgain;
            }
            break;

        case SLASH_STAR:
            *r += c;
            if (c == '*') {
                t_ &= -256;
                t_ |= SLASH_STAR_STAR;
            }
            break;

        case SLASH_STAR_STAR:
            *r += c;
            if (c == '/') {
                *r += HI_RESET;
                goto Pop;
            } else if (c != '*') {
                t_ &= -256;
                t_ |= SLASH_STAR;
            }
            break;

        case QUOTE:
            *r += c;
            if (c == '\'') {
                *r += HI_RESET;
                goto Pop;
            } else if (c == '\\') {
                t_ &= -256;
                t_ |= QUOTE_BACKSLASH;
            }
            break;

        case QUOTE_BACKSLASH:
            *r += c;
            t_ &= -256;
            t_ |= QUOTE;
            break;

        case DQUOTE:
            *r += c;
            if (c == '"') {
                *r += HI_RESET;
                goto Pop;
            } else if (c == '\\') {
                t_ &= -256;
                t_ |= DQUOTE_BACKSLASH;
            }
            break;

        case DQUOTE_BACKSLASH:
            *r += c;
            t_ &= -256;
            t_ |= DQUOTE;
            break;

        Pop:
            t_ >>= 8;
            if (t_ == SELECTOR)
                *r += HI_SELECTOR;
            if (t_ == PROPERTY)
                *r += HI_PROPERTY;
            break;

        default:
            __builtin_unreachable();
        }
    }
}

void
HighlightCss::flush(std::string* r)
{
    while (t_) {
        switch (t_ & 255) {
        case AT:
            *r += '@';
            *r += word_;
            word_.clear();
            break;
        case BANG:
            *r += '!';
            *r += word_;
            word_.clear();
            break;
        case SLASH:
            *r += '/';
            break;
        case SELECTOR:
        case PROPERTY:
        case DQUOTE:
        case DQUOTE_BACKSLASH:
            *r += HI_RESET;
            break;
        default:
            break;
        }
        t_ >>= 8;
    }
    *r += HI_RESET;
    t_ = NORMAL;
}
